Add support for Siri Shortcuts

rmsy opened the pull request:
Hi there 😄 For my Hackweek project, I added support for a few Siri Shortcuts to the Canvas apps. These are the shortcuts I added:

* Check Course Grade - Allows a user to obtain their grade in a course
* Get Courses - Allows a user to obtain a list of courses they are
  enrolled in
* Send Message - Allows a user to send a message to other user(s) using Siri

[Here is my Hackweek video](https://drive.google.com/file/d/1jfRA62baM3UJzdBLFinGZpayHt1Ncjs7/view?usp=sharing), if you're interested in seeing (a brief demo of) how this works.

This is my first contribution to the iOS project, so I'm hopeful that I've done things correctly, but please let me know if there is anything I should improve or do differently. Please also let me know if you have any questions.

Thank you!

**Note**: This change sets some properties to `public` inside of a few API models, in order
to avoid the complexity of interacting with CoreData from within the app
extension (this is the same pattern as what was previously done when
widgets were implemented).

---

refs: LS-2681
affects: Student, Teacher, Parent
release note: Added supoort for interacting with Canvas using Shortcuts and Siri.

test plan:
- in each of the Teacher, Parent, Student apps:
  - log in to the app
  - ask Siri something similar to "send a message using <app>"
    - verify "canvas", "student", "teacher", "parent", "canvas parent"
      are all recognized as app names
  - when prompted, provide a partial name of another user
    - if there are multiple matches, expect to see a disambiguation
      prompt (expect user avatars to be displayed)
    - if there is one match, expect to be asked to confirm the name
  - when prompted, provide message contents, and send the message
  - after sending the message, confirm that tapping on the "chat bubble"
    in the Siri UI opens the newly sent message in the app (Student and
    Teacher apps only)
- in the Student and Teacher apps:
  - verify that the 'Get Courses' action works correctly in the Shortcuts
    app
    - confirm that the filters work, and the results returned by the
      action are valid
  - verify that the 'Check Course Grade' action works correctly in the
    Shortcuts app
    - confirm that selecting a course manually works (list loads, fuzzy
      search works)
    - confirm that providing a course as a value from the 'Get Courses'
      action works
    - confirm that the returned results are valid

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/instructure/canvas-ios?pullRequest=1973) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/instructure/canvas-ios?pullRequest=1973) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/instructure/canvas-ios?pullRequest=1973) it.</sub>

inst-danger commented:
<!--
  1 failure:  Build failed, ski...
  0 warning: 
  
  3 markdown notices
  DangerID: danger-id-default;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Build failed, skipping coverage check
  </td>
    </tr>
  </tbody>
</table>



<details><summary>:x: Swift lint</summary>


    ❌ /Users/vagrant/git/Core/Core/Conversations/APIConversation.swift:283:1: Line Length Violation: Line should be 200 characters or less: currently 208 characters (line_length)
    /Users/vagrant/git/Core/Core/IntentHandlers/CanvasIntentHandler.swift:23:37: error: Redundant Void Return Violation: Returning Void in a function declaration is redundant. (redundant_void_return)
    /Users/vagrant/git/Core/Core/IntentHandlers/SendMessageIntentHandler.swift:68:42: error: Force Cast Violation: Force casts should be avoided. (force_cast)
    ❌ /Users/vagrant/git/Core/Core/IntentHandlers/SendMessageIntentHandler.swift:47:1: Line Length Violation: Line should be 200 characters or less: currently 278 characters (line_length)
    /Users/vagrant/git/Core/Core/IntentHandlers/SendMessageIntentHandler.swift:42:41: error: Redundant Optional Initialization Violation: Initializing an optional variable with nil is redundant. (redundant_optional_initialization)
    /Users/vagrant/git/Core/Core/IntentHandlers/SendMessageIntentHandler.swift:46:1: error: Trailing Whitespace Violation: Lines should not have trailing whitespace. (trailing_whitespace)
    ❌ /Users/vagrant/git/Core/Core/IntentHandlers/SendMessageIntentHandler.swift:88:1: Trailing Whitespace Violation: Lines should not have trailing whitespace. (trailing_whitespace)
    /Users/vagrant/git/Core/Core/IntentHandlers/SendMessageIntentHandler.swift:100:1: error: Trailing Whitespace Violation: Lines should not have trailing whitespace. (trailing_whitespace)
    /Users/vagrant/git/Core/Core/IntentHandlers/SendMessageIntentHandler.swift:104:1: error: Trailing Whitespace Violation: Lines should not have trailing whitespace. (trailing_whitespace)
    ❌ /Users/vagrant/git/Core/Core/IntentHandlers/SendMessageIntentHandler.swift:40:53: Unused Closure Parameter Violation: Unused parameter "response" in a closure should be replaced with _. (unused_closure_parameter)
    /Users/vagrant/git/Core/Core/IntentHandlers/SendMessageIntentHandler.swift:40:63: error: Unused Closure Parameter Violation: Unused parameter "error" in a closure should be replaced with _. (unused_closure_parameter)
    /Users/vagrant/git/Core/Core/IntentHandlers/SendMessageIntentHandler.swift:105:49: error: Unused Closure Parameter Violation: Unused parameter "response" in a closure should be replaced with _. (unused_closure_parameter)
    ❌ /Users/vagrant/git/Core/Core/IntentHandlers/GetCoursesIntentHandler.swift:89:1: Line Length Violation: Line should be 200 characters or less: currently 214 characters (line_length)
    /Users/vagrant/git/Core/Core/IntentHandlers/GetCoursesIntentHandler.swift:23:1: error: Trailing Whitespace Violation: Lines should not have trailing whitespace. (trailing_whitespace)
    /Users/vagrant/git/Core/Core/IntentHandlers/GetCoursesIntentHandler.swift:27:1: error: Trailing Whitespace Violation: Lines should not have trailing whitespace. (trailing_whitespace)
    ❌ /Users/vagrant/git/Core/Core/IntentHandlers/GetCoursesIntentHandler.swift:31:1: Trailing Whitespace Violation: Lines should not have trailing whitespace. (trailing_whitespace)
    /Users/vagrant/git/Core/Core/IntentHandlers/GetCoursesIntentHandler.swift:73:1: error: Trailing Whitespace Violation: Lines should not have trailing whitespace. (trailing_whitespace)
    /Users/vagrant/git/Core/Core/IntentHandlers/GetCoursesIntentHandler.swift:106:1: error: Trailing Whitespace Violation: Lines should not have trailing whitespace. (trailing_whitespace)
    ❌ /Users/vagrant/git/Core/Core/IntentHandlers/CheckCourseGradeIntentHandler.swift:57:1: Trailing Whitespace Violation: Lines should not have trailing whitespace. (trailing_whitespace)
    /Users/vagrant/git/Core/Core/IntentHandlers/CheckCourseGradeIntentHandler.swift:89:1: error: Trailing Whitespace Violation: Lines should not have trailing whitespace. (trailing_whitespace)
    /Users/vagrant/git/Core/Core/IntentHandlers/CheckCourseGradeIntentHandler.swift:104:1: error: Trailing Whitespace Violation: Lines should not have trailing whitespace. (trailing_whitespace)
    ❌ /Users/vagrant/git/Core/Core/IntentHandlers/CheckCourseGradeIntentHandler.swift:125:1: Trailing Whitespace Violation: Lines should not have trailing whitespace. (trailing_whitespace)
    /Users/vagrant/git/Core/Core/IntentHandlers/CheckCourseGradeIntentHandler.swift:151:1: error: Trailing Whitespace Violation: Lines should not have trailing whitespace. (trailing_whitespace)
    /Users/vagrant/git/Core/Core/IntentHandlers/CheckCourseGradeIntentHandler.swift:48:49: error: Unused Closure Parameter Violation: Unused parameter "response" in a closure should be replaced with _. (unused_closure_parameter)
    ❌ /Users/vagrant/git/Core/Core/IntentHandlers/CheckCourseGradeIntentHandler.swift:69:49: Unused Closure Parameter Violation: Unused parameter "response" in a closure should be replaced with _. (unused_closure_parameter)
    /Users/vagrant/git/Core/Core/IntentHandlers/CheckCourseGradeIntentHandler.swift:57:1: error: Vertical Whitespace Violation: Limit vertical whitespace to a single empty line. Currently 2. (vertical_whitespace)
    /Users/vagrant/git/Parent/ParentSiriExtension/IntentHandler.swift:33:1: error: Trailing Whitespace Violation: Lines should not have trailing whitespace. (trailing_whitespace)
</details>


#### Affected Apps: Student, Teacher, Parent

#### Release Note: 
Added supoort for interacting with Canvas using Shortcuts and Siri.
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

vargaat commented:
Hi @rmsy,

Thank you for your contribution. I've seen your video on the hackweek video viewing session, it's really awesome! Currently we are busy with our regular stuff but as soon as we have spare time we'll take deeper look on the code, at first glance I've found no major issues. In the meantime I also activated our test builds to check for any errors and warnings.

rmsy commented:
Hi @vargaat,

Thank you for the reply! I've gone ahead and added a commit to address the issues raised by the linter.

rmsy commented on Core/Core/IntentHandlers/SendMessageIntentHandler.swift:
```diff
@@ -0,0 +1,123 @@
+
+        for (index, request) in requests.enumerated() {
+            env.api.makeRequest(request) { results, _, error in
+                guard error == nil else { return [] }
```
This is returning `[]` from the closure, not from the outer function. Is this still OK? (It compiles fine)

vargaat commented on Core/Core/Courses/APICourse.swift:
```diff
@@ -225,6 +225,22 @@ public struct GetCoursesRequest: APIRequestable {
+        .total_scores,
+    ]
+
+    var include: [Include] = defaultIncludes
```
Yes, sorry. I don't know why I wrote `lines` :)

rmsy commented:
Thank you for the very detailed feedback, @vargaat! I believe I've addressed all of your comments. As for iOS 13, this works OK on iOS 13, as far as I can tell. In the case of both of the warnings you show in your screenshot, they are annotated with the below:
```swift
    @available(iOSApplicationExtension 14.0, *)
```

And the iOS 13 alternative (if it exists; which is only in one of the two cases) is annotated with this:
```swift
    @available(iOSApplicationExtension, introduced: 13.0, deprecated: 14.0, message: "")
```

Is this the correct way to do this? Maybe there's something I'm missing here.

I've verified that both of the actions available through the Shortcuts app work on iOS 13. I was unable to verify the full flow of message sending in the iOS 13 simulator, but I believe that's a problem with Siri on my simulator device. (I can't even check the weather).

vargaat commented:
@rmsy Would it be okay with you if we would merge this into a feature branch of our repository to finalise it? This way it would be more easier for us to address any issues we encounter before merging to master.

rmsy commented:
> @rmsy Would it be okay with you if we would merge this into a feature branch of our repository to finalise it? This way it would be more easier for us to address any issues we encounter before merging to master.

Of course! Feel free 🙂
